nectar/downloaders/threaded.py

```diff
@@ -273,119 +273,128 @@ def _fetch(self, request, session):
+                    chunks = self.chunk_generator(response.raw, self.buffer_size)
+                else:
+                    chunks = response.iter_content(self.buffer_size)
+
+                for chunk in chunks:
+                    if self.is_canceled or request.canceled:
+                        raise DownloadCancelled(request.url)
+
+                    file_handle.write(chunk)
+
+                    bytes_read = len(chunk)
+                    report.bytes_downloaded += bytes_read
+
+                    now = datetime.datetime.now()
+
+                    if now - last_update_time >= progress_interval:
+                        last_update_time = now
+                        self.fire_download_progress(report)
+
+                    with self._bytes_lock:
+                        if now - self._time_bytes_this_second_was_cleared >= ONE_SECOND:
+                            self._bytes_this_second = 0
+                            self._time_bytes_this_second_was_cleared = now
+                        self._bytes_this_second += bytes_read
+
+                    if max_speed is not None and self._bytes_this_second >= max_speed:
+                        # it's not worth doing fancier mathematics than this, very
+                        # fine-grained sleep times [1] are not honored by the system
+                        # [1] for example, sleeping the remaining fraction of time
+                        # before this second is up
+                        time.sleep(0.5)
+
+                # guarantee 1 report at the end
+                self.fire_download_progress(report)
+
+            except SkipLocation:
+                _logger.debug("Skipping {url} because {netloc} could not be reached.".format(
+                    url=request.url, netloc=netloc)
+                )
+                report.download_skipped()
+
+            except requests.ConnectionError as e:
+                if nretry < DEFAULT_TRIES - 1:
+                    continue
+                _logger.error(_('Skipping requests to {netloc} due to repeated connection'
+                                ' failures: {e}').format(netloc=netloc, e=str(e)))
+                self.failed_netlocs.add(netloc)
+                report.download_connection_error()
+
+            except requests.Timeout:
+                if nretry < DEFAULT_TRIES - 1:
+                    continue
+                """
+                Handle a timeout differently than a connection error. Do not add
+                to failed_netlocs so that a new connection can be attempted.
+                """
+                _logger.warning("Request Timeout - Connection with {url} timed out.".format(
+                    url=request.url)
+                )
+                report.download_connection_error()
+
+            except DownloadCancelled as e:
+                _logger.info(str(e))
+                report.download_canceled()
+
+            except DownloadFailed as e:
```

Comment:
@midnightercz Can you please follow the @hao-yu suggestion and re-try here only when there is not response code? e.args[1]? I am suggesting to focus to handle only RST connection error, which basically will be caught in this branch.
I am also leaning towards to decrease the number of tries from 5 to 3.

I would remove the re-try for any other exception, including generic exception, timeout and connection error because it goes against the fail-fast logic.

Reply:
+1 to 3 retries also. In cases where the server situation is unhealthy, so many retries could slow-roll a failure to be very long.